File: go_practice/linked_list.go
package main

import "fmt"

type Node struct {
  next *Node

  val interface{}
}

type List struct {
  head *Node

  count int
}

func (L *List) Prepend(val interface{}) {
  node := &Node{
    next: L.head,
    val: val,
  }

  L.head = node

  l := L.head
  for l.next != nil {
    l = l.next
  }
  L.count++
}

func (L *List) Append(val interface{}) {
  node := &Node{
    val: val,
  }

  if L.head == nil {
    L.head = node
  } else {
    l := L.head
    for l.next != nil {
      l = l.next
    }
    l.next = node
  }
  L.count++
}

func (L *List) Insert(position int, val interface{}) {
  node := &Node{
    val: val,
  }

  l := L.head
  for i := 1; i < position; i++ {
    l = l.next
  }
  holder := l.next
  l.next = node
  l.next.next = holder

  L.count++
}

func (L *List) Pop() interface{} {
  if L.head.next == nil {
    holder := L.head.val
    L.head = nil
    return holder
  } else {
    l := L.head
    for l.next.next != nil {
      l = l.next
    }
    value := l.next.val
    l.next = nil
    L.count--
    return value
  }
}

func (L *List) Shift() interface{} {
  holder := L.head.next
  value := L.head.val
  L.head = holder
  L.count--
  return value
}

func (L *List) Count() int{
  return L.count
}

func (L *List) Show() {
  list := L.head
  for list != nil {
    fmt.Printf("%v -> ", list.val)
    list = list.next
  }
  fmt.Println()
}

func main() {
  l := List{}
  l.Append(10)
  l.Append(10)
  l.Append(10)

  l.Show()
  l.Prepend(12)
  l.Prepend(13)
  l.Prepend(14)

  l.Show()

  l.Append(15)
  l.Show()

  l.Insert(3, 20)
  l.Insert(3, 22)
  l.Show()

  fmt.Println(l.Pop())
  l.Show()

  fmt.Println(l.Shift())
  l.Show()

  fmt.Printf("list: %v\n", l)
  fmt.Printf("count: %v\n", l.Count())
}

File: java_practice/linked_list/LinkedList.java
package linked_list;

public class LinkedList {

  public Node head;
  public int count;

  public LinkedList() {
    count = 0;
  }

  public void append(int data) {
    if (head == null) {
      head = new Node(data);
    } else {
      Node currentNode = head;
      while (currentNode.next != null) {
        currentNode = currentNode.next;
      }
      currentNode.next = new Node(data);
    }
    count++;
  }

  public void prepend(int data) {
    Node placeholder = head;
    head = new Node(data);
    head.next = placeholder;
    count++;
  }

  public void insert(int index, int data) {
    Node currentNode = head;
    for (int i = 1; i <= index; i++) {
      currentNode = currentNode.next;
    }
    Node placeholder = currentNode.next;
    currentNode.next = new Node(data);
    currentNode.next.setNext(placeholder);
    count++;
  }

  public int pop() {
    Node currentNode = head;
    if (head.next != null) {
      while (currentNode.next.next != null) {
        currentNode = currentNode.next;
      }
      int rn = currentNode.next.data;
      currentNode.next = null;
      count--;
      return rn;
    } else {
      int returnedNode = head.next.data;
      head.next = null;
      count--;
      return returnedNode;
    }
  }

  public int shift() {
    int returnedData = head.data;
    head = head.next;
    count--;
    return returnedData;
  }

  public int count() {
    return count;
  }

  public void show() {
    Node currentNode = head;
    if (head != null) {
      for (int i = 0; i < count; i++) {
        System.out.println(currentNode.data);
        currentNode = currentNode.next;
      }
    } else {
      System.out.println("Linked List Empty");
    }
    System.out.println("------------------");
  }

}
